### PR TITLE
feat(scalasdk): Public API for Views, #381

### PR DESCRIPTION
**Pull request by patriknw**
* also an attempt of using adapters to/from javasdk

Refs #381

### CODE REVIEW

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/view/ViewAdapter.scala [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import com.akkaserverless.javasdk.ServiceCallFactory
+import com.akkaserverless.scalasdk.Metadata
+import com.akkaserverless.scalasdk.impl.MetadataImpl
+import com.akkaserverless.scalasdk.view.UpdateContext
+import com.akkaserverless.scalasdk.view.View
+import com.akkaserverless.scalasdk.view.ViewCreationContext
+import com.akkaserverless.scalasdk.view.ViewProvider
+import com.google.protobuf.Descriptors
+
+private[scalasdk] class Scala2JavaViewAdapter[S](scalasdkView: View[S]) extends javasdk.view.View[S] {
```
Comment (patriknw) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/view/ViewAdapter.scala:33]: This adapter approach is completely untested and I'm not sure it's the right approach.

Comment (patriknw) [reply sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/view/ViewAdapter.scala:33]: This is the "flow":

1. AkkaServerless.register(scalasdk.ViewProvider)
  - delegates to javasdk by wrapping the provider with Scala2JavaViewProviderAdapter
2. javasdk.ViewProvider.newHandler is invoked
  - adapter delegates to scaladsl new handler and adapts the ViewCreationContext
  - the action View implementation is created
3. For each incoming event
  - javasdk.ViewHandler._internalSetUpdateContext  is invoked
    - adapter delegates to scaladsl and adapts the UpdateContext
  - javasdk.ViewHandler.handleUpdate is invoked
    - adapter delegates to scaladsl
    - adapter converts the returned effect

Comment (johanandren) [reply sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/view/ViewAdapter.scala:33]: A bit of wrapping/unwrapping work but sounds reasonable.

**Review by johanandren — COMMENTED** [review]
Looks ok to me

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/view/View.scala [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+   */
+  protected final def updateContext(): UpdateContext =
+    _updateContext.getOrElse(
+      throw new IllegalStateException("UpdateContext is only available when handling an update."))
+
+  /** INTERNAL API */
+  private[scalasdk] def _internalSetUpdateContext(context: Option[UpdateContext]): Unit =
+    _updateContext = context
+
+  protected final def effects(): View.UpdateEffect.Builder[S] =
```
Comment (octonato) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/view/View.scala:89]: shall we remove the `()` on the methods?

Comment (patriknw) [reply sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/view/View.scala:89]: yes, that might read be better

**Review by raboof — APPROVED** [review]
*Looks* good AFAICS so far ;)

**Review by octonato — APPROVED** [review]
I think there are a few more methods with `()`, but we can revisit it later on in an API polishing round.

**Comment by patriknw** [comment]
There is also `updateContext`, but for that one I think it's good to have it as `updateContext()` since it changes for each call.

In the EffectBuilder we have `ignore()`. That could be argued, but I kind of like having the `()` in the effect builder even though it might be ok without for immutable builders like that.

**Pull request merged**